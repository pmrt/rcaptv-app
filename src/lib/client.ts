import { QueryClient } from "@tanstack/react-query";

// Time in Ms
const Sec = 1000;
const Min = Sec * 60;

export const client = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 10 * Min,
      retry: false,
    },
  },
});
