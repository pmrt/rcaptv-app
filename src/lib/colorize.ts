const interpColor = (rgbA: number[], rgbB: number[], f: number) => {
  return rgbA
    .slice()
    .map((rgb, i) => Math.round(rgb + f * (rgbB[i] - rgbA[i])));
};

// const outliers = (sorted: number[]) => {
//   const q1 = sorted[Math.floor(sorted.length / 4)];
//   const q3 = sorted[Math.floor((3 * sorted.length) / 4)];
//   const iqr = q3 - q1;
//   return [q1 - 1.5 * iqr, q3 + 1.5 * iqr];
// };

export const colorize = (rgbA: number[], rgbB: number[], arr: number[]) => {
  const sorted = arr.sort((a, b) => a - b);
  // const [lower, upper] = outliers(sorted);
  // const filtered = sorted.filter((n) => n >= lower && n <= upper);
  return (idx: number) =>
    interpColor(rgbA, rgbB, (1 / (sorted.length - 1)) * idx);
};
