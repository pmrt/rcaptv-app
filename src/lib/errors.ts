export class RcapError extends Error {
  wrappedErr: Error | undefined;
  name = "RcapError";

  err(err: Error): this {
    this.wrappedErr = err;
    return this;
  }

  string() {
    return `${this.name}: ${this.message}`;
  }
}

export class ErrPageNotFound extends RcapError {
  name = "ErrPageNotFound";
}

export class ErrInvalidStreamer extends RcapError {
  name = "ErrInvalidStreamer";
}

export class ErrResponse extends RcapError {
  name = "ErrResponse";
}

type RcapResponse = {
  data: unknown;
  errors: string[];
};
export class ErrFetch extends RcapError {
  name = "ErrFetch";
  status = 0;
  statusText = "";
  errors: string[] = [];

  // Reads asynchronously from the response to extract error info
  async asyncResp(r: Response) {
    this.status = r.status;
    this.statusText = r.statusText;
    const parsed: RcapResponse = await r.json();
    if (parsed?.errors?.length > 0) {
      this.errors = parsed.errors;
    }
    return this;
  }

  string() {
    return this.status
      ? `${this.name}: ${this.message} (${this.status} ${this.statusText})`
      : `${this.name}: ${this.message}`;
  }
}

export function isError(value: unknown): value is Error {
  return value instanceof Error;
}
