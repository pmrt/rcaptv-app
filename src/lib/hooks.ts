import { useEffect, useState } from "react";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";

import type { AppDispatch, RootState } from "@/store";

export const useExternalScript = (src: string) => {
  const [isReady, setIsReady] = useState(false);

  useEffect(() => {
    const script = document.createElement("script");

    script.src = src;
    script.defer = true;

    document.body.appendChild(script);

    const onLoad = () => {
      setIsReady(true);
    };
    script.addEventListener("load", onLoad);
    return () => {
      document.body.removeChild(script);
      script.removeEventListener("load", onLoad);
    };
  }, [src]);

  return isReady;
};

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
