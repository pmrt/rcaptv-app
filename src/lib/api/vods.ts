import { fetchOrFail, getURL } from "@/lib/api/api";
import { ErrFetch } from "@/lib/errors";

type VOD = {
  id: string;
  user_id: string;
  stream_id: string;
  created_at: string;
  published_at: string;
  language: string;
  title: string;
  thumbnail_url: string;
  view_count: number;
  duration_seconds: number;
};
type VODResponse = {
  data: {
    vods: VOD[];
  };
  errors: string[];
};
export const getVodsByStreamer = async (
  username: string
): Promise<VODResponse> => {
  const resp = await fetchOrFail(getURL("/vods", { username }));
  if (!resp.ok) {
    throw await new ErrFetch("").asyncResp(resp);
  }
  return resp.json();
};
