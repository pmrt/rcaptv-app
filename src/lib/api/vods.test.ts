import { ErrFetch } from "@/lib/errors";
import { getVodsByStreamer } from "./vods";

describe("api:vods", () => {
  test("getVods", async () => {
    const vods = await getVodsByStreamer("knekro");
    expect(vods.data.vods).toMatchSnapshot();
  });

  test("getVods:emptyBid", async () => {
    try {
      const vods = await getVodsByStreamer("");
      console.log(vods);
    } catch (e) {
      expect(e).toBeInstanceOf(ErrFetch);
      expect((e as ErrFetch).errors).toHaveLength(1);
      expect((e as ErrFetch).errors[0]).toBe("Missing username or vid");
    }
  });
});
