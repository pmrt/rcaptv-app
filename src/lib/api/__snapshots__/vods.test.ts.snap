// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api:vods getVods 1`] = `
[
  {
    "created_at": "2023-06-27T16:39:55Z",
    "duration_seconds": 13337,
    "id": "1857217060",
    "language": "es",
    "published_at": "2023-06-27T16:39:55Z",
    "stream_id": "46980798892",
    "thumbnail_url": "https://vod-secure.twitch.tv/_404/404_processing_%{width}x%{height}.png",
    "title": "Tu maravilloso dia de variedad !eneba",
    "user_id": "152633332",
    "view_count": 514,
  },
  {
    "created_at": "2023-06-26T15:07:30Z",
    "duration_seconds": 34740,
    "id": "1856303227",
    "language": "es",
    "published_at": "2023-06-26T15:07:30Z",
    "stream_id": "46977536044",
    "thumbnail_url": "https://static-cdn.jtvnw.net/cf_vods/dgeft87wbj63p/3207b5118aab280a2c7d_knekro_46977536044_1687792045//thumb/thumb0-%{width}x%{height}.jpg",
    "title": "ONLY CUAJADO| Empezamos la semana sin tilteos  | !eneba",
    "user_id": "152633332",
    "view_count": 213601,
  },
  {
    "created_at": "2023-06-25T19:06:43Z",
    "duration_seconds": 16204,
    "id": "1855660030",
    "language": "es",
    "published_at": "2023-06-25T19:06:43Z",
    "stream_id": "46975624444",
    "thumbnail_url": "https://static-cdn.jtvnw.net/cf_vods/dgeft87wbj63p/a62ab40050be44efcaed_knekro_46975624444_1687719997//thumb/thumb0-%{width}x%{height}.jpg",
    "title": "DOMINGUITO PERRON DE VARIEDAD | !eneba",
    "user_id": "152633332",
    "view_count": 106342,
  },
  {
    "created_at": "2023-06-24T18:30:08Z",
    "duration_seconds": 22838,
    "id": "1854762977",
    "language": "es",
    "published_at": "2023-06-24T18:30:08Z",
    "stream_id": "46972379804",
    "thumbnail_url": "https://static-cdn.jtvnw.net/cf_vods/dgeft87wbj63p/4683313680d0286bb0ad_knekro_46972379804_1687631403//thumb/thumb0-%{width}x%{height}.jpg",
    "title": "Sábado de variedad y espectáculo  !eneba",
    "user_id": "152633332",
    "view_count": 152643,
  },
  {
    "created_at": "2023-06-23T16:37:41Z",
    "duration_seconds": 31471,
    "id": "1853788795",
    "language": "es",
    "published_at": "2023-06-23T16:37:41Z",
    "stream_id": "46968878348",
    "thumbnail_url": "https://static-cdn.jtvnw.net/cf_vods/dgeft87wbj63p/ad2208cb25e441c66c16_knekro_46968878348_1687538255//thumb/thumb0-%{width}x%{height}.jpg",
    "title": "CHALLENGE STREAM | RETO EN JUEGOS POR SUBS, SE VIENE !eneba",
    "user_id": "152633332",
    "view_count": 217067,
  },
  {
    "created_at": "2023-06-22T17:09:08Z",
    "duration_seconds": 26260,
    "id": "1852942266",
    "language": "es",
    "published_at": "2023-06-22T17:09:08Z",
    "stream_id": "46965345212",
    "thumbnail_url": "https://static-cdn.jtvnw.net/cf_vods/dgeft87wbj63p/4fa29e27adaf0ea0c0ac_knekro_46965345212_1687453743//thumb/thumb0-%{width}x%{height}.jpg",
    "title": "DIa TOP| Variedad + Radio KOI 21:00 + 7º Aniversario DBD (?) | Dia weon | !eneba",
    "user_id": "152633332",
    "view_count": 112487,
  },
]
`;
